### PR TITLE
Added the UserLogListener.java class

### PR DESCRIPTION
**Pull request by davidolorundare**
See [JENKINS-54087](https://issues.jenkins-ci.org/browse/JENKINS-54087).

When I tried compiling the UserLogListener it gave errors that it couldn't find the `hudson.Extension` and `hudson.ExtensionList` classes. The other updates to the `catalog.json` are within this pull request too though.

Desired Reviewers:
@jvz

### CODE REVIEW

File: src/main/java/io/jenkins/plugins/audit/RequestContext.java [hunk 1]
```diff
@@ -39,7 +39,7 @@
     @ClientServer
     public static final String REQUEST_URI = "requestUri";
     @ClientServer
-    public static final String NODE_NAME = "nodeName";
+    public static final String NODE_NAME = "master";
```
Comment (jvz) [inline src/main/java/io/jenkins/plugins/audit/RequestContext.java]: This should be the attribute name, not the attribute value. We'll be setting the value to `master` by default, though.

**Comment by jvz** [comment]
You should rebase this from the master branch. I squashed your previous PR, so now we have diverging git histories.

File: src/main/java/io/jenkins/plugins/audit/RequestContext.java (hunk 1 above)
Comment (davidolorundare) [reply src/main/java/io/jenkins/plugins/audit/RequestContext.java]: oh I already changed that in the last commit. .. or wait.. you mean replacing 'NODE_NAME' with just 'MASTER' ?

File: src/main/resources/catalog.json [hunk 2]
```diff
@@ -10,9 +10,6 @@
     "attributes" : [ {
       "name" : "timestamp",
       "required" : true
-    }, {
-      "name" : "requestId",
```
Comment (jvz) [inline src/main/resources/catalog.json:14]: Why are you removing this?

File: src/main/java/io/jenkins/plugins/audit/listeners/UserLogListener.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    Logout logout = LogEventFactory.getEvent(Logout.class);
+    String hostName = NetUtils.getLocalHostname();
+    String inetAddress = InetAddress.getLocalHost().getHostAddress();
+
+    /**
+     * Fired when a user has logged in, event logged via Log4j-audit.
+     * @param username name or ID of the user who logged in.
+     */
+    @Override
+    public static void fireLoggedIn(@Nonnull String username) {
```
Comment (jvz) [inline src/main/java/io/jenkins/plugins/audit/listeners/UserLogListener.java]: You can't override a static method.

File: src/main/java/io/jenkins/plugins/audit/listeners/UserLogListener.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+import org.apache.logging.log4j.audit.LogEventFactory;
+import org.apache.logging.log4j.core.util.NetUtils;
+
+/**
+ * Listener notified of user login and logout events.
+ */
+ @Extension
+public class UserLogListener extends SecurityListener {
+
+    Login login = LogEventFactory.getEvent(Login.class);
```
Comment (jvz) [inline src/main/java/io/jenkins/plugins/audit/listeners/UserLogListener.java]: I don't think any of these fields should be fields. Shared instances of data classes across threads is not a good idea without synchronization, and I'm not so sure if we really need synchronization here if we just create these values inside the methods.

File: src/main/java/io/jenkins/plugins/audit/RequestContext.java [hunk 5]
```diff
@@ -80,7 +79,7 @@ public static String getRequestId() {
             uuid = UuidUtil.getTimeBasedUuid();
             ThreadContext.put(REQUEST_ID, uuid.toString());
         }
-        return uuidStr;
+        return ThreadContext.get(REQUEST_ID);
```
Comment (jvz) [inline src/main/java/io/jenkins/plugins/audit/RequestContext.java]: This change is unnecessary. `uuidStr` already contains the updated value.

**Review by jvz — CHANGES_REQUESTED** [review]

File: src/main/resources/catalog.json (hunk 2 above)
Comment (davidolorundare) [reply src/main/resources/catalog.json:14]: `requestId` is a general attribute.  ... i removed the `requestId` and `requestUri` attributes from the events and made them be passed as `RequestContext` attributes, as per the [doc](https://logging.apache.org/log4j-audit/latest/requestContext.html)

File: src/main/java/io/jenkins/plugins/audit/RequestContext.java (hunk 5 above)
Comment (davidolorundare) [reply src/main/java/io/jenkins/plugins/audit/RequestContext.java]: but if `uuidStr` is null then null will be returned

Comment (jvz) [reply src/main/java/io/jenkins/plugins/audit/RequestContext.java]: Oh, good point. Then maybe you could reassign uuidStr inside the if block. You don't even need the UUID variable; you can convert it straight to a string right away.

File: src/main/resources/catalog.json (hunk 2 above)
Comment (jvz) [reply src/main/resources/catalog.json:14]: Oh got it.

File: src/main/java/io/jenkins/plugins/audit/listeners/UserLogListener.java (hunk 3 above)
Comment (davidolorundare) [reply src/main/java/io/jenkins/plugins/audit/listeners/UserLogListener.java]: lol... forgot. will do method hiding for that and fix it in next commit

File: src/main/java/io/jenkins/plugins/audit/listeners/UserLogListener.java (hunk 4 above)
Comment (davidolorundare) [reply src/main/java/io/jenkins/plugins/audit/listeners/UserLogListener.java]: alright, i'll put them inside the methods then.

File: src/main/java/io/jenkins/plugins/audit/listeners/UserLogListener.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+     *
+     * UserLogListener does not use this abstract class method, however, rather than leaving
+     * its implementation empty it is made to throw an exception if the method is ever called.
+     *
+     * @param details details of the newly authenticated user, such as name and groups.
+     * @throws UnsupportedOperationException if the method is called wrongly.
+     */
+    @Override
+    protected void authenticated(@Nonnull UserDetails details) {
+        throw new UnsupportedOperationException("This class does not use this method.");
```
Comment (jvz) [inline src/main/java/io/jenkins/plugins/audit/listeners/UserLogListener.java]: You should just make these methods no-ops. That way, when this listener is invoked, it doesn't cause any errors elsewhere. These methods can be safely ignored.

File: src/main/java/io/jenkins/plugins/audit/listeners/UserLogListener.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+     * @param username name or ID of the user who logged in.
+     */
+    @Override
+    protected void loggedIn(@Nonnull String username) {
+        String currentTime = Instant.now().toString();
+        Login login = LogEventFactory.getEvent(Login.class);
+
+        try {
+            RequestContext.setHostName(NetUtils.getLocalHostname());
+            RequestContext.setIpAddress(InetAddress.getLocalHost().getHostAddress());
```
Comment (jvz) [inline src/main/java/io/jenkins/plugins/audit/listeners/UserLogListener.java]: This is not the correct IP address. We want the requester IP address. For example: `Stapler.getCurrentRequest().getRemoteAddr()`

File: src/main/java/io/jenkins/plugins/audit/listeners/UserLogListener.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+     *
+     * @param username name or ID of the user who logged in.
+     */
+    @Override
+    protected void loggedIn(@Nonnull String username) {
+        String currentTime = Instant.now().toString();
+        Login login = LogEventFactory.getEvent(Login.class);
+
+        try {
+            RequestContext.setHostName(NetUtils.getLocalHostname());
```
Comment (jvz) [inline src/main/java/io/jenkins/plugins/audit/listeners/UserLogListener.java]: We don't have a hostName attribute anywhere. Can you remove this from the RequestContext?

**Review by jvz — CHANGES_REQUESTED** [review]

File: src/main/java/io/jenkins/plugins/audit/listeners/UserLogListener.java (hunk 6 above)
Comment (davidolorundare) [reply src/main/java/io/jenkins/plugins/audit/listeners/UserLogListener.java]: alright

**Review by jvz — APPROVED** [review]
Looks good to me!

**Pull request merged**